File: src/services/tauri/langgraph.ts
import { invoke } from '@tauri-apps/api/core';

export interface LLMRequest {
  prompt: string;
  model?: string;
  temperature?: number;
  max_tokens?: number;
}

export interface LLMResponse {
  content: string;
  model: string;
  tokens_used?: number;
}

export interface WorkflowState {
  id: string;
  project_id: string;
  current_step: string;
  data: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface WorkflowStep {
  name: string;
  description: string;
  tool_name: string;
  inputs: Record<string, any>;
  outputs: Record<string, any>;
  next_steps: string[];
  completed: boolean;
}

export interface WorkflowDefinition {
  name: string;
  description: string;
  steps: WorkflowStep[];
  initial_step: string;
}

export class LangGraphService {
  // Core LangGraph tools
  static async callLLM(request: LLMRequest): Promise<LLMResponse> {
    try {
      return await invoke('call_llm', { request });
    } catch (error) {
      console.error('Failed to call LLM:', error);
      throw error;
    }
  }

  static async queryLocalSQLite(query: string, params: string[] = []): Promise<any[]> {
    try {
      const result = await invoke('query_local_sqlite', { query, params });
      return JSON.parse(result as string);
    } catch (error) {
      console.error('Failed to query local SQLite:', error);
      throw error;
    }
  }

  static async getProblemContext(problemId?: string): Promise<any> {
    try {
      const result = await invoke('get_problem_context', { problemId });
      return JSON.parse(result as string);
    } catch (error) {
      console.error('Failed to get problem context:', error);
      throw error;
    }
  }

  static async saveGenerationResult(
    problemId: string,
    resultType: string,
    content: string,
    metadata?: string
  ): Promise<void> {
    try {
      await invoke('save_generation_result', {
        problemId,
        resultType,
        content,
        metadata
      });
    } catch (error) {
      console.error('Failed to save generation result:', error);
      throw error;
    }
  }

  static async getAvailableTools(): Promise<any[]> {
    try {
      return await invoke('get_available_tools');
    } catch (error) {
      console.error('Failed to get available tools:', error);
      throw error;
    }
  }

  // LangGraph orchestrator methods
  static async executeWorkflow(
    workflowName: string,
    projectId: string,
    initialData: Record<string, any>
  ): Promise<WorkflowState> {
    try {
      const result = await invoke('execute_langgraph_workflow', {
        workflowName,
        projectId,
        initialData
      });
      return result as WorkflowState;
    } catch (error) {
      console.error('Failed to execute workflow:', error);
      throw error;
    }
  }

  static async listWorkflows(): Promise<string[]> {
    try {
      return await invoke('list_langgraph_workflows');
    } catch (error) {
      console.error('Failed to list workflows:', error);
      throw error;
    }
  }

  static async getWorkflowDefinition(workflowName: string): Promise<WorkflowDefinition> {
    try {
      const result = await invoke('get_langgraph_workflow_definition', {
        workflowName
      });
      return result as WorkflowDefinition;
    } catch (error) {
      console.error('Failed to get workflow definition:', error);
      throw error;
    }
  }

  // High-level workflow methods
  static async runProblemToSolutionWorkflow(
    projectId: string,
    problemInput: string
  ): Promise<WorkflowState> {
    const initialData = {
      problem_input: problemInput
    };

    return this.executeWorkflow('problem_to_solution', projectId, initialData);
  }

  static async validateProblem(problemInput: string): Promise<{
    validated_problem: string;
    is_valid: boolean;
    feedback: string;
  }> {
    const request: LLMRequest = {
      prompt: `Please analyze this problem statement and provide a refined, validated version:

${problemInput}

Provide:
1. A clear, refined problem statement
2. Whether the problem is valid and solvable
3. Any recommendations for improvement

Format your response as JSON with keys: validated_problem, is_valid, feedback`,
      model: 'gemini-pro',
      temperature: 0.7,
      max_tokens: 1024
    };

    try {
      const response = await this.callLLM(request);
      return JSON.parse(response.content);
    } catch (error) {
      console.error('Failed to validate problem:', error);
      throw error;
    }
  }

  static async generatePersonas(validatedProblem: string): Promise<Array<{
    name: string;
    demographics: string;
    role: string;
    goals: string[];
    pain_points: string[];
    tech_level: string;
  }>> {
    const request: LLMRequest = {
      prompt: `Based on this problem: ${validatedProblem}

Generate 3-5 detailed user personas who would be affected by this problem. For each persona, provide:
1. Name and basic demographics
2. Role/occupation
3. Goals and motivations
4. Pain points related to this problem
5. Technical proficiency level

Format as JSON array with objects containing: name, demographics, role, goals, pain_points, tech_level`,
      model: 'gemini-pro',
      temperature: 0.8,
      max_tokens: 2048
    };

    try {
      const response = await this.callLLM(request);
      return JSON.parse(response.content);
    } catch (error) {
      console.error('Failed to generate personas:', error);
      throw error;
    }
  }

  static async generateSolutions(
    validatedProblem: string,
    personas: any[]
  ): Promise<Array<{
    name: string;
    description: string;
    target_personas: string[];
    technical_approach: string;
    business_impact: string;
    complexity: number;
  }>> {
    const request: LLMRequest = {
      prompt: `Problem: ${validatedProblem}

Personas: ${JSON.stringify(personas)}

Generate 3-5 key solutions that address this problem for these personas. For each solution:
1. Solution name
2. Description
3. Which personas it serves
4. Technical approach
5. Business impact
6. Implementation complexity (1-10)

Format as JSON array with objects containing: name, description, target_personas, technical_approach, business_impact, complexity`,
      model: 'gemini-pro',
      temperature: 0.8,
      max_tokens: 2048
    };

    try {
      const response = await this.callLLM(request);
      return JSON.parse(response.content);
    } catch (error) {
      console.error('Failed to generate solutions:', error);
      throw error;
    }
  }

  // Utility methods
  static async testConnection(): Promise<boolean> {
    try {
      const tools = await this.getAvailableTools();
      return tools.length > 0;
    } catch (error) {
      console.error('LangGraph connection test failed:', error);
      return false;
    }
  }

  static async healthCheck(): Promise<{
    llm_available: boolean;
    database_available: boolean;
    workflows_available: boolean;
  }> {
    try {
      const [llmTest, dbTest, workflowsTest] = await Promise.allSettled([
        this.callLLM({ prompt: 'test', model: 'gemini-pro' }),
        this.queryLocalSQLite('SELECT 1'),
        this.listWorkflows()
      ]);

      return {
        llm_available: llmTest.status === 'fulfilled',
        database_available: dbTest.status === 'fulfilled',
        workflows_available: workflowsTest.status === 'fulfilled'
      };
    } catch (error) {
      console.error('Health check failed:', error);
      return {
        llm_available: false,
        database_available: false,
        workflows_available: false
      };
    }
  }
} 

File: src-tauri/src/tools/langgraph.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;
use chrono::{DateTime, Utc};
use anyhow::Result;

use super::{LLMRequest, LLMResponse, call_llm, query_local_sqlite, get_problem_context, save_generation_result};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowState {
    pub id: String,
    pub project_id: String,
    pub current_step: String,
    pub data: HashMap<String, serde_json::Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStep {
    pub name: String,
    pub description: String,
    pub tool_name: String,
    pub inputs: HashMap<String, serde_json::Value>,
    pub outputs: HashMap<String, serde_json::Value>,
    pub next_steps: Vec<String>,
    pub completed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowDefinition {
    pub name: String,
    pub description: String,
    pub steps: Vec<WorkflowStep>,
    pub initial_step: String,
}

pub struct LangGraphOrchestrator {
    pub workflow_definitions: HashMap<String, WorkflowDefinition>,
}

impl LangGraphOrchestrator {
    pub fn new() -> Self {
        let mut orchestrator = Self {
            workflow_definitions: HashMap::new(),
        };
        
        orchestrator.register_default_workflows();
        orchestrator
    }
    
    fn register_default_workflows(&mut self) {
        // Register the problem-to-solution workflow
        let problem_workflow = WorkflowDefinition {
            name: "problem_to_solution".to_string(),
            description: "Complete workflow from problem validation to solution generation".to_string(),
            steps: vec![
                WorkflowStep {
                    name: "validate_problem".to_string(),
                    description: "Validate and refine the problem statement".to_string(),
                    tool_name: "call_llm".to_string(),
                    inputs: HashMap::from([
                        ("prompt_template".to_string(), serde_json::json!(
                            "Please analyze this problem statement and provide a refined, validated version:\n\n{problem_input}\n\nProvide:\n1. A clear, refined problem statement\n2. Whether the problem is valid and solvable\n3. Any recommendations for improvement\n\nFormat your response as JSON with keys: validated_problem, is_valid, feedback"
                        ))
                    ]),
                    outputs: HashMap::new(),
                    next_steps: vec!["generate_personas".to_string()],
                    completed: false,
                },
                WorkflowStep {
                    name: "generate_personas".to_string(),
                    description: "Generate user personas based on the validated problem".to_string(),
                    tool_name: "call_llm".to_string(),
                    inputs: HashMap::from([
                        ("prompt_template".to_string(), serde_json::json!(
                            "Based on this problem: {validated_problem}\n\nGenerate 3-5 detailed user personas who would be affected by this problem. For each persona, provide:\n1. Name and basic demographics\n2. Role/occupation\n3. Goals and motivations\n4. Pain points related to this problem\n5. Technical proficiency level\n\nFormat as JSON array with objects containing: name, demographics, role, goals, pain_points, tech_level"
                        ))
                    ]),
                    outputs: HashMap::new(),
                    next_steps: vec!["generate_solutions".to_string()],
                    completed: false,
                },
                WorkflowStep {
                    name: "generate_solutions".to_string(),
                    description: "Generate solutions based on personas and problem".to_string(),
                    tool_name: "call_llm".to_string(),
                    inputs: HashMap::from([
                        ("prompt_template".to_string(), serde_json::json!(
                            "Problem: {validated_problem}\n\nPersonas: {personas}\n\nGenerate 3-5 key solutions that address this problem for these personas. For each solution:\n1. Solution name\n2. Description\n3. Which personas it serves\n4. Technical approach\n5. Business impact\n6. Implementation complexity (1-10)\n\nFormat as JSON array with objects containing: name, description, target_personas, technical_approach, business_impact, complexity"
                        ))
                    ]),
                    outputs: HashMap::new(),
                    next_steps: vec!["complete".to_string()],
                    completed: false,
                },
            ],
            initial_step: "validate_problem".to_string(),
        };
        
        self.workflow_definitions.insert("problem_to_solution".to_string(), problem_workflow);
    }
    
    pub async fn execute_workflow(&self, workflow_name: &str, project_id: &str, initial_data: HashMap<String, serde_json::Value>) -> Result<WorkflowState> {
        let workflow = self.workflow_definitions.get(workflow_name)
            .ok_or_else(|| anyhow::anyhow!("Workflow {} not found", workflow_name))?;
        
        let mut state = WorkflowState {
            id: Uuid::new_v4().to_string(),
            project_id: project_id.to_string(),
            current_step: workflow.initial_step.clone(),
            data: initial_data,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        };
        
        // Execute steps in sequence
        let mut current_step_name = workflow.initial_step.clone();
        while current_step_name != "complete" {
            let step = workflow.steps.iter()
                .find(|s| s.name == current_step_name)
                .ok_or_else(|| anyhow::anyhow!("Step {} not found", current_step_name))?;
            
            state = self.execute_step(step, state).await?;
            
            // Move to next step (simplified - just take first next step)
            current_step_name = step.next_steps.first()
                .unwrap_or(&"complete".to_string())
                .clone();
        }
        
        Ok(state)
    }
    
    async fn execute_step(&self, step: &WorkflowStep, mut state: WorkflowState) -> Result<WorkflowState> {
        state.current_step = step.name.clone();
        state.updated_at = Utc::now();
        
        match step.tool_name.as_str() {
            "call_llm" => {
                // Build prompt from template and state data
                let prompt_template = step.inputs.get("prompt_template")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                
                let prompt = self.build_prompt(prompt_template, &state.data)?;
                
                let llm_request = LLMRequest {
                    prompt,
                    model: Some("gemini-pro".to_string()),
                    temperature: Some(0.7),
                    max_tokens: Some(2048),
                };
                
                // This is a simplified call - in real implementation you'd need proper async handling
                // For now, this shows the structure
                let response = self.call_llm_tool(llm_request).await?;
                
                // Parse and store response
                let result: serde_json::Value = serde_json::from_str(&response.content)
                    .unwrap_or_else(|_| serde_json::json!({"content": response.content}));
                
                state.data.insert(step.name.clone(), result.clone());
                
                // Save to database
                self.save_step_result(&state.project_id, &step.name, &result).await?;
            }
            "query_local_sqlite" => {
                // Execute database query
                let query = step.inputs.get("query")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                
                let params = step.inputs.get("params")
                    .and_then(|v| v.as_array())
                    .unwrap_or(&vec![])
                    .iter()
                    .map(|v| v.to_string())
                    .collect();
                
                let result = self.query_database(query.to_string(), params).await?;
                state.data.insert(step.name.clone(), result);
            }
            _ => {
                return Err(anyhow::anyhow!("Unknown tool: {}", step.tool_name));
            }
        }
        
        Ok(state)
    }
    
    fn build_prompt(&self, template: &str, data: &HashMap<String, serde_json::Value>) -> Result<String> {
        let mut prompt = template.to_string();
        
        for (key, value) in data {
            let placeholder = format!("{{{}}}", key);
            let replacement = match value {
                serde_json::Value::String(s) => s.clone(),
                _ => value.to_string(),
            };
            prompt = prompt.replace(&placeholder, &replacement);
        }
        
        Ok(prompt)
    }
    
    async fn call_llm_tool(&self, request: LLMRequest) -> Result<LLMResponse> {
        // In a real implementation, this would call the actual tool
        // For now, simulate the response
        Ok(LLMResponse {
            content: "{}".to_string(),
            model: request.model.unwrap_or_else(|| "gemini-pro".to_string()),
            tokens_used: Some(100),
        })
    }
    
    async fn query_database(&self, query: String, params: Vec<String>) -> Result<serde_json::Value> {
        // In a real implementation, this would call the actual database tool
        // For now, return empty result
        Ok(serde_json::json!([]))
    }
    
    async fn save_step_result(&self, project_id: &str, step_name: &str, result: &serde_json::Value) -> Result<()> {
        // In a real implementation, this would call the actual save tool
        // For now, just log
        log::info!("Saving step {} result for project {}", step_name, project_id);
        Ok(())
    }
    
    pub fn get_workflow_status(&self, workflow_name: &str) -> Option<&WorkflowDefinition> {
        self.workflow_definitions.get(workflow_name)
    }
    
    pub fn list_workflows(&self) -> Vec<String> {
        self.workflow_definitions.keys().cloned().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_orchestrator_creation() {
        let orchestrator = LangGraphOrchestrator::new();
        assert!(orchestrator.workflow_definitions.contains_key("problem_to_solution"));
    }
    
    #[test]
    fn test_prompt_building() {
        let orchestrator = LangGraphOrchestrator::new();
        let template = "Hello {name}, your problem is: {problem}";
        let mut data = HashMap::new();
        data.insert("name".to_string(), serde_json::json!("John"));
        data.insert("problem".to_string(), serde_json::json!("test problem"));
        
        let result = orchestrator.build_prompt(template, &data).unwrap();
        assert_eq!(result, "Hello John, your problem is: test problem");
    }
} 